### ISSUE TITLE
jnp.log/log2/log10 return -inf for subnormal float32 inputs (XLA DAZ flush)

### ISSUE THREAD

**Issue by seanonymoussubmission** [jax-ml/jax#37670]
### Description

`jnp.log`, `jnp.log2`, and `jnp.log10` return `-inf` for subnormal `float32` inputs on the CPU backend, while NumPy returns the correct finite negative values.

This is a downstream consequence of XLA's DAZ (Denormals-Are-Zero) flush: subnormal inputs are flushed to `0.0` before the log operation, resulting in `log(0) = -inf`. While the [Sharp Bits docs](https://docs.jax.dev/en/latest/notebooks/Common_Gotchas_in_JAX.html) mention that subnormal semantics differ from NumPy, the specific `-inf` outcome for log functions is not documented and can silently corrupt computations.

```python
import warnings, jax, jax.numpy as jnp, numpy as np

tiny = np.finfo(np.float32).tiny
xs = np.array([np.nextafter(tiny, np.float32(0)), 1e-38, 5e-39, 1.4e-45], dtype=np.float32)

for name, np_op, jnp_op in [("log", np.log, jnp.log), ("log2", np.log2, jnp.log2), ("log10", np.log10, jnp.log10)]:
    with warnings.catch_warnings():
        warnings.simplefilter("ignore")
        expected = np_op(xs)
    actual = np.asarray(jnp_op(jnp.array(xs)))
    print(f"jnp.{name}: expected={expected}  actual={actual}")
```

Output:

`jnp.log:   expected=[ -87.33655  -87.49824  -88.19138 -103.27893]  actual=[-inf -inf -inf -inf]`
`jnp.log2:  expected=[-126.      -126.23327 -127.23327 -149.     ]  actual=[-inf -inf -inf -inf]`
`jnp.log10: expected=[-37.92978 -38.      -38.30103 -44.85347]      actual=[-inf -inf -inf -inf]`

The inputs are valid representable subnormal floats. In gradient-based workflows, a silent `-inf` propagates as `nan` and can crash training with no obvious cause. A documentation clarification — or ideally a runtime warning when a log-family function receives an input flushed to zero — would prevent silent failures.

### System info (python version, jaxlib version, accelerator, etc.)

jax        0.10.1.dev20260513+64f6e229b
jaxlib     0.10.0
Python     3.12.3
Backend    cpu

Note: tested on a nightly build. Given that this relates to documented DAZ/XLA behavior, it likely reproduces on stable as well.

**Comment by jakevdp** [comment]
I think this is expected: in general JAX (via XLA) uses flush-to-zero semantics for subnormals; see https://docs.jax.dev/en/latest/notebooks/Common_Gotchas_in_JAX.html#miscellaneous-divergences-from-numpy for a mention of this.

Because `log(0)` is `-inf`, the outputs you're seeing above are what we would expect.

> A documentation clarification — or ideally a runtime warning when a log-family function receives an input flushed to zero — would prevent silent failures.

It's documented in sharp bits; I think it would be overkill to re-document this behavior in every function's doc string.

Regarding a runtime warning, that's difficult to do without severe performance penalties. For operations on accelerators, there is no efficient mechanism to propagate warnings back to the host that are conditional on values that appear within arrays at runtime.

**Comment by jakevdp** [comment]
xref #38651
